### PR TITLE
🐛 Destination GCS: CMEK is supported for the connector

### PR DESCRIPTION
**Pull request by armsepehr**
## What
*Describe what the change is solving*
This branch may fix the following issue: GCS destination is not supported Customer-Managed Encryption Key and it is mentioned in the readme file for the destination. 
1. https://github.com/airbytehq/airbyte/issues/18195
2. https://github.com/airbytehq/airbyte/pull/18315

The config file has a missing field which is misleading me for several hour, so I have included the field and some sample-files to check the destination commands as well as the integration tests easily.

## How
*Describe the solution*
As it is discussed in the issue there are two main methods to solve these issues: First, rewrite GCS without AmazonS3 base class. Second, may set integrity check only for GCS to false. In this branch, the second solution is selected. 

Note that at the current implementation, google-cloud-storage has been connected to the server via amazon s3 classes and they have some difference.

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

The change guarantees that the users can send their data to the gcs bucket which is enable by customer-managed encryption key feature. 

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [x] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

### CODE REVIEW

**Comment by marcosmarxm** [comment]
@armsepehr please add the checklist you removed from the template and fill the tasks done. 
@armsepehr please sign the CLA.

**Comment by armsepehr** [comment]
@marcosmarxm 
Should I do anything more?

File: airbyte-integrations/connectors/destination-gcs/sample_files/configured_catalog.json [hunk 1]
```diff
@@ -0,0 +1,27 @@
+{
+  "streams": [
+    {
+      "sync_mode": "full_refresh",
+      "destination_sync_mode": "append",
+      "stream": {
+        "name": "ab-airbyte-testing",
+        "supported_sync_modes": ["full_refresh"],
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/destination-gcs/sample_files/configured_catalog.json:8]: Why are you adding this file?

File: airbyte-integrations/connectors/destination-gcs/sample_secrets/config.json [hunk 2]
```diff
@@ -6,5 +6,8 @@
     "credential_type": "HMAC_KEY",
     "hmac_key_access_id": "<access-id>",
     "hmac_key_secret": "<secret>"
+  },
+  "format": {
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/destination-gcs/sample_secrets/config.json:10]: This is not related to this change.

File: airbyte-integrations/connectors/destination-gcs/sample_files/messages.jsonl [hunk 3]
```diff
@@ -0,0 +1,2 @@
+{"type": "RECORD", "record": {"stream": "ab-airbyte-testing", "data": {"_ab_pk": "my_value", "column2": 221, "column3": "2021-01-01T20:10:22", "column4": 1.214, "column5": [1,2,3]}, "emitted_at": 1626172757000}}
+{"type": "RECORD", "record": {"stream": "ab-airbyte-testing", "data": {"_ab_pk": "my_value2", "column2": 222, "column3": "2021-01-02T22:10:22", "column5": [1,2,null]}, "emitted_at": 1626172757000}}
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/destination-gcs/sample_files/messages.jsonl:2]: Same for this file.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/destination-gcs/sample_files/configured_catalog.json (hunk 1 above)
Comment (armsepehr) [reply airbyte-integrations/connectors/destination-gcs/sample_files/configured_catalog.json:8]: This file is necessary to verify the "write" command locally. I see the same file in other connector, so I included in gcs too.

File: airbyte-integrations/connectors/destination-gcs/sample_secrets/config.json (hunk 2 above)
Comment (armsepehr) [reply airbyte-integrations/connectors/destination-gcs/sample_secrets/config.json:10]: This is not related to this change, but I think the creator missed to include the format in "config.json" file. I waste some time to figure out the sample "config.json" has not the format field which is necessary. It give us some strange error message if you do not include the "format" field. In a word, I think it is better to include "format" field in "config.json" file.

File: airbyte-integrations/connectors/destination-gcs/sample_files/messages.jsonl (hunk 3 above)
Comment (armsepehr) [reply airbyte-integrations/connectors/destination-gcs/sample_files/messages.jsonl:2]: As I said before, this file was included in other connectors and I think it's necessary to verify the connector locally. In my opinion, it is better to include in the project same as other connector.

**Comment by marcosmarxm** [comment]
Hello :wave::skin-tone-2: and thank you for your contribution!

Airbyte has instituted a code freeze between 19 and 30 December, to make sure there are no disruptions during the holidays. 
Because of this, reviewing and merging your contribution may take longer than usual. 
We apologize for the delay, but we want everyone to have a quiet and happy holiday.

If you have any questions or need further clarification, please don't hesitate to ping via Slack.

**Comment by grishick** [comment]
Rebased, bumped version, and running tests here: https://github.com/airbytehq/airbyte/pull/21682
